pkg/ddevapp/config_merge.go

```diff
@@ -0,0 +1,206 @@
+package ddevapp
+
+import (
+	"errors"
+	"log"
+	"regexp"
+
+	"github.com/imdario/mergo"
+	"gopkg.in/yaml.v2"
+)
+
+// mergeConfigToApp does an unmarshall with merging
+func (app *DdevApp) mergeConfigToApp(source []byte) error {
+	// save away state before merges.
+	unmergedApp := &DdevApp{}
+	err := mergo.Merge(unmergedApp, app)
+	if err != nil {
+		return err
+	}
+
+	type mergeData struct {
+		newData interface{}
+		oldData interface{}
+	}
+
+	// get the updated settings. Note that we will replace
+	// anything else from the upstream config for any
+	// key except for web_environment.
+	err = yaml.Unmarshal(source, app)
+	if err != nil {
+		return err
+	}
+
+	// add merges here.  Default strategy is to clobber old
+	// keys.
+	mergeableItems := map[string]mergeData{
```

Comment:
As above, I think mergo should do *all* of this for you, without any need for any of this?

Reply:
Open to that. Not hard, but as we've seen, it's not straightforward to merge structs, which don't act like most languages. So each struct needs to be handled a la carte.  

That list tells you what I'm processing currently.  I think the one thing I skipped is the database struct, which I can add to this PR if you like.